core/src/main/java/bio/terra/pearl/core/service/participant/merge/ParticipantMergeService.java

```diff
@@ -0,0 +1,199 @@
+        this.kitRequestService = kitRequestService;
+        this.participantUserService = participantUserService;
+        this.mergeDao = mergeDao;
+    }
+
+    @Transactional
+    public ParticipantUserMerge applyMerge(ParticipantUserMerge merge, DataAuditInfo auditInfo) {
+        for (MergeAction<Enrollee, EnrolleeMerge> enrolleeMerge : merge.getEnrollees()) {
+            applyMergeEnrollee(enrolleeMerge, merge, auditInfo);
+        }
+        portalParticipantUserService.delete(merge.getPpUsers().getSource().getId(), CascadeProperty.EMPTY_SET);
+        participantUserService.delete(merge.getUsers().getSource().getId(), CascadeProperty.EMPTY_SET);
+        return merge;
+    }
+
+    protected void applyMergeEnrollee(MergeAction<Enrollee, EnrolleeMerge> mergeAction, ParticipantUserMerge parentMerge, DataAuditInfo auditInfo) {
+        if (mergeAction.getAction().equals(MergeAction.Action.NO_ACTION)) {
+            // nothing to do
+        } else if (mergeAction.getAction().equals(MergeAction.Action.DELETE_SOURCE)) {
+            // delete the source enrollee
+            deleteMergedEnrollee(mergeAction.getSource(), auditInfo);
+        } else if (mergeAction.getAction().equals(MergeAction.Action.MOVE_SOURCE)) {
+            // move the source enrollee to the target user
+            moveEnrollee(mergeAction.getSource(), parentMerge.getPpUsers().getTarget(), auditInfo);
+        } else if (mergeAction.getAction().equals(MergeAction.Action.MERGE)) {
+            // merge the source enrollee into the target enrollee
+            mergeEnrolleeData(mergeAction.getSource(), mergeAction.getTarget(), mergeAction.getMergePlan(), parentMerge.getPpUsers().getTarget(), auditInfo);
+            deleteMergedEnrollee(mergeAction.getSource(), auditInfo);
+        }
+    }
+
+    protected void deleteMergedEnrollee(Enrollee enrollee, DataAuditInfo auditInfo) {
+        withdrawnEnrolleeService.withdrawEnrollee(enrollee, auditInfo);
+    }
+
+    /** merges data from one enrollee into another.  Does not delete the source enrollee */
+    protected void mergeEnrolleeData(Enrollee source, Enrollee target, EnrolleeMerge mergePlan, PortalParticipantUser targetPpUser, DataAuditInfo auditInfo) {
+        for (MergeAction<ParticipantTask,?> action : mergePlan.getTasks()) {
+            if (action.getAction().equals(MergeAction.Action.DELETE_SOURCE)) {
+                deleteTask(action.getSource(), auditInfo);
+            } else if (action.getAction().equals(MergeAction.Action.MOVE_SOURCE) ||
+                    action.getAction().equals(MergeAction.Action.MERGE) ||
+                    action.getAction().equals(MergeAction.Action.MOVE_SOURCE_DELETE_TARGET)) {
+                moveTask(action.getSource().getId(), target, targetPpUser, auditInfo);
+                if (action.getAction().equals(MergeAction.Action.MOVE_SOURCE_DELETE_TARGET)) {
+                    deleteTask(action.getTarget(), auditInfo);
+                }
```

Comment:
Do we need to also delete the target on merge? Otherwise we'll end up with two tasks right?

Reply:
Eventually, yes.  But since we're not merging the SurveyResponses (yet), we have to keep both tasks